DRILL-7562: Support HTTP Basic authentication for REST API calls

dobesv opened the pull request:
This can greatly simplify the development of HTTP clients, as well as
use from the command line using curl/wget, since you don't have to
deal with storing the session cookie.

# [DRILL-7562](https://issues.apache.org/jira/browse/DRILL-7562): Support HTTP Basic authentication for REST API calls

## Description

Implement support for Basic authentication.  This allows an HTTP client to supply the login credentials in the same request as the query.  For clients that do not wish to keep track of cookies, this saves an HTTP round-trip per query.

This is useful for access via the command-line, but also for client libraries where storing the cookie is a nuisance.

## Documentation

----------

# Authenticating REST API requests

## Basic authentication

Basic authentication refers to the standard HTTP authentication method of the same name.  Most HTTP clients and tools provide a simple way to do Basic authentication.

Basic authentication is enabled or disabled using `drill-override.conf`.  Add the string `"BASIC"` to `http.auth.mechanisms`.  Note that if the field is not currently set, it defaults to having `"FORM"` in it, so you probably want to include `"FORM"` if you set this field, so that Web UI users can still log in.

Example:

```
http: {
    enabled: true,
    auth: {
        # Http Auth mechanisms to configure. If not provided but user.auth is enabled
        # then default value is ["FORM"].
        mechanisms: ["BASIC", "FORM"]
    }
}
```

To authenticate requests using Basic authentication, send the appropriate `Authorization` header with each request using your HTTP client's options:

    curl -kv \
           -u drilluser:drillpassword  \
           -X POST \
           -H "Content-Type: application/json" \
           -d '{"queryType":"SQL", "query": "select * from sys.version"}' \
           http://localhost:8047/query.json

## Form based authentication

Form-based authentication is enabled or disabled using `drill-override.conf`.  Add the string `"FORM"` to `http.auth.mechanisms` if it is set.  If `http.auth.mechanisms` is not set, `"FORM"` is enabled by default.

Example:

```
http: {
    enabled: true,
    auth: {
        # Http Auth mechanisms to configure. If not provided but user.auth is enabled
        # then default value is ["FORM"].
        mechanisms: ["BASIC", "FORM"]
    }
}
```

To authenticate requests using form-based authentication, you must use an HTTP client that saves cookies between requests.  Simulate a form submission to the same URL used in the Web UI / Console (`/j_security_check`)

    curl -X POST \
        -H "Content-Type: application/x-www-form-urlencoded" \
        -k -c cookies.txt -s \
        -d "j_username=drilluser" \
        -d "j_password=drillpassword" \
        http://localhost:8047/j_security_check


In subsequent requests, use the cookie returned from that request:

    curl -kv \
           -b cookies.txt  \
           -X POST \
           -H "Content-Type: application/json" \
           -d '{"queryType":"SQL", "query": "select * from sys.version"}' \
           http://localhost:8047/query.json


---------

## Testing

I tested it using `curl -u user:pass "localhost:8049/storage.json"`.  When the feature was active, the request would succeed.  If not, the server returned a redirect to the login page.

Assistance in writing a unit test would be appreciated, I didn't see one for the current login system to base this off and I didn't want to spend a ton of time trying to figure out how to write a test for this one.

cgivre commented:
@dobesv Thanks for the contribution and Welcome to the Drill community!  
I had a few minor formatting comments.  Drill has a formatting doc for both Eclipse and Intellij which is available here[1].  Please run those on the new file and make sure that the formatting is consistent with the Drill style. 

Once those are committed, I think this will get committed pretty quickly. 
Thanks again!

[1]: https://drill.apache.org/docs/apache-drill-contribution-guidelines/

dobesv commented:
@cgivre Thanks for the review!  I've applied the formatting rules you linked and added the missing copyright header to the new file.  Hopefully this does the trick.

dobesv commented:
I just updated the commit message to include the JIRA issue.

ihuzenko commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/auth/DrillHttpSecurityHandlerProvider.java:
```diff
@@ -191,6 +200,7 @@ public boolean isFormEnabled() {
             AuthStringUtil.asSet(config.getStringList(ExecConstants.HTTP_AUTHENTICATION_MECHANISMS)));
       } else {
         // For backward compatibility
+        configuredMechs.add(Constraint.__BASIC_AUTH);
```
Please don't enable basic by default.

ihuzenko commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/auth/DrillHttpSecurityHandlerProvider.java:
```diff
@@ -139,10 +140,14 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
         securityHandler.handle(target, baseRequest, request, response);
+      } else if(isBasicEnabled() && request.getHeader(HttpHeader.AUTHORIZATION.asString()) != null) {
+        final DrillHttpConstraintSecurityHandler basicSecurityHandler = securityHandlers.get(Constraint.__BASIC_AUTH);
+        basicSecurityHandler.handle(target, baseRequest, request, response);
```
please reuse already defined ```securityHandler``` variable as was done for other cases.

ihuzenko commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/auth/HttpBasicAuthSecurityHandler.java:
```diff
@@ -0,0 +1,48 @@
+      throw new DrillException("BASIC auth mechanism was configured but PLAIN mechanism is not enabled to provide an " +
+        "authenticator. Please configure user authentication with PLAIN mechanism and authenticator to use " +
+        "BASIC authentication");
+    }
```
this is very similar to check done in  ```FormSecurityHanlder``` , maybe makes sense to create protected method in base class and use it for both cases?

dobesv replied:
I don't think it's worth creating whole class just to reuse what amounts to basically 2 lines of code.  Maybe something for another day.

ihuzenko commented:
Hello @dobesv , it looks like there is no well-developed framework for REST tests, but I think for this purpose you can extend _**RestClientFixture**_ class and try using it in your test.

dobesv commented:
@ihuzenko I think writing tests is going to be too much for me to do on this, I've already spent more time on this PR than I should have.

Not only would I need to spin up a REST server fixture, but I would need it to have some kind of username & password authentication system setup with pre-defined usernames and password.  If you would like to help me by adding some tests to this PR that would be great! Otherwise I think I  lack the project-specific skill needed to do that in a reasonable amount of time.

dobesv commented:
I updated the PR description to include documentation for both the Form-based and Basic authentication mechanisms.

cgivre commented:
@ihuzenko 
I'd argue that writing a test framework for the REST features is out of scope for this PR.  We should probably create a JIRA for that separately, but I'd vote for committing this as long as we are comfortable that it isn't causing any regressions.

dobesv commented:
@ihuzenko Actually I hadn't noticed these classes shared a base class, I had thought they only implemented an interface for some reason.  I refactored that check up into the base class.  However, I haven't figured out a way to confirm that the message is logged correctly.  When I mess around with the configuration I can never actually see that message anywhere.  Not that it is likely to be super important for the message to be correct, if it's not the stack trace will give all the information anyone needs.  But if you know a way to test that message, let me know.

I also fixed a bug I noticed while I was working on that, and resolved conflicts.

dobesv commented:
I put the documentation into the PR description - I guess there is some process that copies the documentation to the site after a release?

dobesv commented:
@ihuzenko I have rebased and squashed, please let me know if there's anything more I can do to help!

Thank you for your quick turnaround on the PRs!

dobesv commented:
I created a docs PR here, if it's useful:

https://github.com/apache/drill/pull/1980
